components/Buttons/src/ColorThemer/MDCButton+MDCStyle.h

```diff
@@ -0,0 +1,56 @@
+/*
+ Copyright 2017-present the Material Components for iOS authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+#import "MaterialButtons.h"
+
+@protocol MDCColorScheme;
+
+@interface MDCButton (MDCStyle)
+
+/**
+ Returns a new MDCButton styled as a "default size" floating action button. If @c colorScheme is
+ provided, it will be applied to the button. The button is 56 points in diameter and supports a
+ 24-point square icon.
+
+ @param colorScheme The color scheme to apply. May be nil.
+ @return An MDCButton styled as a floating action button with the color scheme applied.
+ */
++ (nonnull MDCButton *)floatingButtonWithScheme:(nullable NSObject<MDCColorScheme> *)colorScheme;
+
+/**
+ Returns a new MDCButton styled as a "mini size" floating action button. If @c colorScheme is
+ provided, it will be applied to the button. The button is 40 points in diameter and supports a
+ 24-point square icon.
+
+ @param colorScheme The color scheme to apply. May be nil.
+ @return An MDCButton styled as a floating action button with the color scheme applied.
+ */
++ (nonnull MDCButton *)miniFloatingButtonWithScheme:
+    (nullable NSObject<MDCColorScheme> *)colorScheme;
+
+/**
+ Returns a new MDCButton styled as a "default size" floating action button with a large (36 point)
+ icon. If @c colorScheme is provided, it will be applied to the button. The button is 56 points in
+ diameter and supports a 24-point square icon.
+
+ @param colorScheme The color scheme to apply. May be nil.
+ @return An MDCButton styled as a floating action button with the color scheme applied.
+ */
++ (nonnull MDCButton *)largeIconFloatingButtonWithScheme:
```

Comment:
At this point I wager whether it's worth simply aligning with the UIKit API convention of `buttonWithType`.

```objc
+ (nonnull MDCButton *)buttonWithMaterialType:(MDCButtonType);
+ (nonnull MDCButton *)buttonWithMaterialType:(MDCButtonType)
                                  colorScheme:(nullable NSObject<MDCColorScheme> *)colorScheme;
+ (nonnull MDCButton *)buttonWithType:(UIButtonType) NS_UNAVAILABLE;
```

Reply:
If you go this route, then you could also make the buttonType property settable in order to transform the type at runtime.

Reply:
Mmmm I just want to say that I'm not a fan of the UIKit `buttonWithType` API so I'd be super cautious of adopting it just for the sake of alignment with UIKit here.

Reply:
What are the caveats that you're concerned about?

Reply:
I agree with @jverkoey  that it feels very similar, but I'm on the side of not doing enumerations. However, I can't come up with a good technical argument right now.  Since it's contentious and a convenience category I think it's better to drop the category from this PR and just focus on the actual styler/themer/typer/mutator APIs until later.

Reply:
Most of it is a gut feeling but tying ourselves to an enum of button types is IMO a bad idea. I feel our styles should be more separated from the implementation than an assignable enum would allow.

Reply:
+1 to dropping the category until a stronger case can be made.

Reply:
> I feel our styles should be more separated from the implementation than an assignable enum would allow.

Do you foresee a future where we'd want to provide additional arguments to the initializer?

Reply:
I see a situation where we might want to perform a chain of "mutations" to the base button to achieve some styled result:

* Transform into one of the Material button "types"
* Update colors to match branding
* Mutate text/imagery to ensure accessibility conformance
* Perform mutation X

I have a (possibly unfounded) concern that by using an enumeration we may back ourselves into a corner of having to support those styles indefinitely. Especially if those styles are very specific.

Reply:
Got it! That's a solid concern.